cpp/src/gcylon/sorting/sorting.cpp

```diff
@@ -0,0 +1,263 @@
+    // we perform sorting on this table, and revert back to the original at the end
+    std::vector<int32_t> column_indices_4s;
+    for (int i = 0; i < tv.num_columns(); ++i) {
+        if (std::find(sort_column_indices.begin(), sort_column_indices.end(), i) == sort_column_indices.end()) {
+            column_indices_4s.push_back(i);
+        }
+    }
+    column_indices_4s.insert(column_indices_4s.begin(), sort_column_indices.begin(), sort_column_indices.end());
+    auto tv_for_sorting = tv.select(column_indices_4s);
+
+    // sort unspecified columns in ASCENDING order
+    std::vector<cudf::order> column_orders_all(tv.num_columns(), cudf::order::ASCENDING);
+    for (int i = 0; i < sort_column_indices.size(); ++i) {
+        column_orders_all[i] = column_orders[i];
+    }
+
+    cudf::null_order null_ordering = nulls_after ? cudf::null_order::AFTER : cudf::null_order::BEFORE;
+    std::vector<cudf::null_order> column_null_orders(tv.num_columns(), null_ordering);
+
+    // first perform local sort
+    auto initial_sorted_tbl = cudf::sort(tv_for_sorting, column_orders_all, column_null_orders);
+    auto initial_sorted_tv = initial_sorted_tbl->view();
+
+    // get sort columns as a separate table_view
+    std::vector<int32_t> key_column_indices;
+    for (int i = 0; i < sort_column_indices.size(); ++i) {
+        key_column_indices.push_back(i);
+    }
+    auto sort_columns_tv = initial_sorted_tbl->select(key_column_indices);
+
+    // sample the sorted table with sort columns and create a table
+    int sample_count = ctx->GetWorldSize() * SAMPLING_RATIO;
+    if (sample_count > sort_columns_tv.num_rows()) {
+        sample_count = sort_columns_tv.num_rows();
+    }
+    std::unique_ptr<cudf::table> sample_tbl;
+    RETURN_CYLON_STATUS_IF_FAILED(
+            gcylon::SampleTableUniform(sort_columns_tv, sample_count, sample_tbl));
+    auto sample_tv = sample_tbl->view();
+
+    // get split points
+    std::unique_ptr<cudf::table> split_points_table;
+    RETURN_CYLON_STATUS_IF_FAILED(
+            gcylon::GetSplitPoints(sample_tv,
+                                   sort_root,
+                                   column_orders,
+                                   ctx,
+                                   split_points_table));
+    auto split_points_tv = split_points_table->view();
+
+    // calculate split indices on the initial sorted table
+    std::vector<int32_t> split_point_indices;
+    RETURN_CYLON_STATUS_IF_FAILED(
+            GetSplitPointIndices(sort_columns_tv,
+                                 split_points_tv,
+                                 column_orders,
+                                 null_ordering,
+                                 split_point_indices));
+
+    // perform all to all on split_points
+    std::vector<std::unique_ptr<cudf::table>> received_tables;
+    // add first and the last row index to the split indices
+    split_point_indices.insert(split_point_indices.begin(), 0);
```

Comment:
this insert at the front makes a memcpy on the other elements, and its not elegant. Can we start off by inserting this 0 to `split_point_indices`?

Reply:
that would be kind of ugly, because we are getting the split points through mpi broadcast.